### PR TITLE
fix(python): parse PEP 695 'type' alias statements in the menhir parser

### PR DESCRIPTION
**Pull request by timols**
`type X = expr` (PEP 695, Python 3.12+) was a hard syntax error in the menhir parser — two bare `NAME` tokens in a row don't reduce under any `small_stmt` production.

## Implementation

- **Parsing_hacks_python.ml**: rewrite `NAME("type")` to a dedicated `TYPE` token, same soft-keyword approach as match/case (#665) — only when it unambiguously heads a `type X = ...` statement, so `type(x)`, `type = 5`, etc. are untouched.
- **Parser_python.mly**: `TYPE` token + `type_alias_stmt` rule.
- **AST_python.ml**: new `TypeAlias` node.
- **Python_to_generic.ml**: converts to `DefStmt (ent, TypeDef { tbody = AliasType _ })`, same as Go/C++/Dart's type aliases.

No generics (`type X[T] = ...`) yet — still fails to parse, same as before.

## Tests

**Parsing regression** (`tests/parsing/python/`):
- `type_alias.py` — all supported forms: basic aliases, complex RHS (`Callable`, unions, optionals, nested types), sequential aliases, inside function/class bodies, inline after colon.
- `type_soft_keyword.py` — backward compatibility: `type` as variable, builtin call, function parameter, attribute, keyword argument, expression context, plus mixed usage alternating aliases and identifier use in the same file.

**Semgrep pattern matching** (`tests/patterns/python/`):
- `type_alias_basic` — `type $X = $Y` matches type aliases, does not match `type = 5`, `type(42)`, or `def f(type)`.
- `type_alias_specific` — `type $X = list[int]` matches only the alias with that specific RHS.

## Test plan

- [x] `lang parsing > Python > type_alias.py` — PASS
- [x] `lang parsing > Python > type_soft_keyword.py` — PASS
- [x] `sgrep patterns > Python > type_alias_basic.py` — PASS
- [x] `sgrep patterns > Python > type_alias_specific.py` — PASS

### CODE REVIEW

File: languages/python/menhir/Parsing_hacks_python.ml [hunk 1]
```diff
@@ -167,11 +167,41 @@ let rec rewrite depth match_depths prev_bound = function
...
+let rec rewrite_type_alias_aux prev_bound = function
+  | [] -> []
+  | (T.NAME ("type", ii) as t) :: xs when prev_bound ->
+      if starts_type_alias xs then T.TYPE ii :: rewrite_type_alias_aux false xs
+      else t :: rewrite_type_alias_aux false xs
+  | t :: xs ->
+      let prev_bound' =
+        if is_trivia t then prev_bound else is_stmt_boundary t
+      in
+      t :: rewrite_type_alias_aux prev_bound' xs
```
[the remark this answers is not in the document]
Comment (timols) [reply languages/python/menhir/Parsing_hacks_python.ml]: Good catch, fixed in the latest push: the boundary check now also treats `:` as opening a statement, since `suite` allows a same-line `simple_stmt` after it. Verified `if cond: type Alias = int` now parses, and re-checked dict/slice/lambda/annotation colons don't get misread as alias heads.

**Comment by corneliuhoffman** [comment]
seeing no tests here

**Comment by timols** [comment]
@corneliuhoffman yep, my bad. Sorry - added tests just now.

**Comment by dimitris-m** [comment]
This branch is 210 commits behind main: the interfile merge (#779) and a menhir bump to 20250912 have landed since. Please rebase onto current main, adapt the parser changes, and re-run the Python tests.

This is not a review.

**Comment by timols** [comment]
Hi @dimitris-m I've rebased the changes onto latest main and verified everything still works correctly/no specific changes were needed.

**Comment by timols** [comment]
Hi @dimitris-m @corneliuhoffman @maciejpirog - any chance you could review also this PR and provide feedback? This is also generating lots of parsing errors on our python-based repository.

**Comment by dimitris-m** [comment]
> Hi @dimitris-m @corneliuhoffman @maciejpirog - any chance you could review also this PR and provide feedback? This is also generating lots of parsing errors on our python-based repository.

Hi, we will review it tomorrow.

Be aware that if it is merged now it will go to 2.0 which is the next release.

**Review by corneliuhoffman — COMMENTED** [review]
I think this is quite good, it would benefit from a few more testing.
The one issue I can see is in the range, if done as you are doing it now
```python
type A = int
```
will get as range the range of `A = int`.  This might affect autofix rules

**Comment by dimitris-m** [comment]
please rebase on main also

**Comment by timols** [comment]
Thanks for the review @corneliuhoffman! And thanks for flagging the range issue.

I've fixed the range issue by carrying the token as an `unhandled_keywordattr` on the entity, using the same pattern from elsewhere in the codebase (e.g. Dart’s keyword attrs). The range now correctly spans the full statement.

I also added additional tests: parsing tests for semicolon boundaries, for/while/else inline suites, string literal RHS, ternary expressions, comprehensions, lambda parameters, dict keys; plus a new pattern-matching test for type aliases inside function and class bodies.

While investigating the range fix, I noticed that Go (Go_to_generic.ml:644) and Dart (Parse_dart_tree_sitter.ml:2764) both drop their keyword tokens in AliasType conversions the same way. Happy to create Github issues for those in the project if helpful but didn't want to expand the scope of this PR to other languages.

**Comment by timols** [comment]
@dimitris-m done!

**Comment by dimitris-m** [comment]
CI checks will fail (no-merge-commits), this is why I suggested rebase, we don't merge main on feature branches

**Comment by timols** [comment]
@dimitris-m yep, that's done. I fixed that up. I see CI is fully passing
